searx/engines/duckduckgo.py

```diff
@@ -60,42 +62,30 @@
 __CACHE = []
 
 
-def _cache_key(data: dict):
-    return 'SearXNG_ddg_web_vqd' + redislib.secret_hash(f"{data['q']}//{data['kl']}")
+def _cache_key(query: str, region: str):
+    return 'SearXNG_ddg_web_vqd' + redislib.secret_hash(f"{query}//{region}")
 
 
-def cache_vqd(data: dict, value):
+def cache_vqd(query: str, region: str, value: str):
     """Caches a ``vqd`` value from a query."""
     c = redisdb.client()
     if c:
-        logger.debug("cache vqd value: %s", value)
-        c.set(_cache_key(data), value, ex=600)
+        logger.debug("VALKEY cache vqd value: %s (%s)", value, region)
+        c.set(_cache_key(query, region), value, ex=600)
 
     else:
-        logger.debug("MEM cache vqd value: %s", value)
+        logger.debug("MEM cache vqd value: %s (%s)", value, region)
         if len(__CACHE) > 100:  # cache vqd from last 100 queries
             __CACHE.pop(0)
-        __CACHE.append((_cache_key(data), value))
-
+        __CACHE.append((_cache_key(query, region), value))
 
-def get_vqd(data):
-    """Returns the ``vqd`` that fits to the *query* (``data`` from HTTP POST).
 
-    DDG's bot detection is sensitive to the ``vqd`` value.  For some search terms
-    (such as extremely long search terms that are often sent by bots), no ``vqd``
-    value can be determined.
+def get_vqd(query: str, region: str, force_request: bool = False):
+    """Returns the ``vqd`` that fits to the *query*.
 
-    If SearXNG cannot determine a ``vqd`` value, then no request should go out
-    to DDG:
-
-        A request with a wrong ``vqd`` value leads to DDG temporarily putting
-        SearXNG's IP on a block list.
-
-        Requests from IPs in this block list run into timeouts.
-
-    Not sure, but it seems the block list is a sliding window: to get my IP rid
-    from the bot list I had to cool down my IP for 1h (send no requests from
-    that IP to DDG).
+    :param query: The query term
+    :param region: DDG's region code
+    :param force_request: force a request to get a vqd value from DDG
```

Comment:
Added force_request argument to the function -> so the caching is encapsulated in duckduckgo.py and calls from duckduckgo_extra.py do not need to implement anything for caching.

Reply:
That's a good idea indeed, I thought about doing something similar too, but didn't want to make that function more complex than it is already. Your solution is a nice one though.